File: PitBoard(Iphone)/PitBoardKit/Sources/PitBoardKit/Schedule/Sources/JsonLdSportsEventScheduleSource.swift
import Foundation
import SwiftSoup

/// Fuente genérica para las webs que publican cada ronda con datos estructurados
/// `schema.org/SportsEvent` (bloque `<script type="application/ld+json">` con un array
/// `subEvent`, cada uno con `name` tipo "Practice - Australian Grand Prix" y `startDate`
/// en ISO-8601 con offset) — equivalente exacto de `JsonLdSportsEventScheduleSource.kt`.
/// La reutilizan F2, F3, ELMS y WEC y Le Mans Cup (ver `RaceScheduleRepository.swift`).
///
/// Funciona en dos pasos: 1) se lee la página de listado de la temporada y se sacan los
/// enlaces a cada ronda (empiezan por `roundHrefPrefixTemplate`, con el año ya
/// sustituido); 2) se visita cada ronda y se lee su JSON-LD.
///
/// HONESTO: si el sitio cambia el marcado del listado o dejara de usar JSON-LD, esta
/// fuente empieza a devolver listas vacías (fallo silencioso y aislado, ver
/// `RaceScheduleRepository`) — no hay tabla HTML de la que caer hacia atrás.
public final class JsonLdSportsEventScheduleSource: RaceScheduleSource, @unchecked Sendable {
    public let series: RaceSeries
    private let baseUrl: String
    /// URL de listado de la temporada, con "{year}" como marcador.
    private let listingUrlTemplate: String
    /// Prefijo del href de cada ronda dentro del listado, con "{year}" como marcador.
    private let roundHrefPrefixTemplate: String
    /// Fragmentos que, si aparecen en el slug de la ronda, la descartan (ej. "test").
    private let excludeSlugContaining: [String]

    public init(
        series: RaceSeries,
        baseUrl: String,
        listingUrlTemplate: String,
        roundHrefPrefixTemplate: String,
        excludeSlugContaining: [String] = []
    ) {
        self.series = series
        self.baseUrl = baseUrl
        self.listingUrlTemplate = listingUrlTemplate
        self.roundHrefPrefixTemplate = roundHrefPrefixTemplate
        self.excludeSlugContaining = excludeSlugContaining
    }

    public func fetch() async throws -> [EventDraft] {
        let year = Calendar(identifier: .gregorian).component(.year, from: Date())
        let listingUrl = listingUrlTemplate.replacingOccurrences(of: "{year}", with: String(year))
        let roundHrefPrefix = roundHrefPrefixTemplate.replacingOccurrences(of: "{year}", with: String(year))

        let listingHtml = try await HTTPClient.fetchHTML(listingUrl)
        let listingDoc = try SwiftSoup.parse(listingHtml, baseUrl)

        let anchors = try listingDoc.select("a[href]").array()
        var seen = Set<String>()
        var roundUrls: [String] = []
        for anchor in anchors {
            let href = try anchor.attr("href")
            guard href.hasPrefix(roundHrefPrefix) else { continue }
            guard let absolute = try? anchor.absUrl("href"), !absolute.isEmpty else { continue }
            guard seen.insert(absolute).inserted else { continue }
            let lowered = absolute.lowercased()
            if excludeSlugContaining.contains(where: { lowered.contains($0.lowercased()) }) { continue }
            roundUrls.append(absolute)
        }

        var events: [EventDraft] = []
        for roundUrl in roundUrls {
            if let sessions = try? await sessionsForRound(roundUrl) {
                events.append(contentsOf: sessions)
            }
        }
        return events
    }

    private func sessionsForRound(_ roundUrl: String) async throws -> [EventDraft] {
        let html = try await HTTPClient.fetchHTML(roundUrl)
        let doc = try SwiftSoup.parse(html, roundUrl)

        let scripts = try doc.select("script[type=application/ld+json]").array()
        var event: JsonLdSportsEvent?
        for script in scripts {
            let json = script.data()
            guard let data = json.data(using: .utf8),
                  let decoded = try? JSONDecoder().decode(JsonLdSportsEvent.self, from: data) else { continue }
            if let sub = decoded.subEvent, !sub.isEmpty {
                event = decoded
                break
            }
        }
        guard let event else { return [] }

        let trimmed = roundUrl.hasSuffix("/") ? String(roundUrl.dropLast()) : roundUrl
        let slug = trimmed.components(separatedBy: "/").last ?? trimmed

        // El nombre del subEvent trae "Sesión - Nombre del Gran Premio" — la parte de
        // después del guion es más corta y legible que el "name" del evento raíz (que en
        // fiaformula2/3.com viene con el título completo del GP, patrocinador incluido).
        let subEvents = event.subEvent ?? []
        var roundName = ""
        for sub in subEvents {
            if let name = sub.name, let range = name.range(of: " - ") {
                let candidate = String(name[range.upperBound...]).trimmingCharacters(in: .whitespacesAndNewlines)
                if !candidate.isEmpty {
                    roundName = candidate
                    break
                }
            }
        }
        if roundName.isEmpty {
            roundName = event.name?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
        }
        let circuitName = event.location?.name?.trimmingCharacters(in: .whitespacesAndNewlines) ?? ""

        var result: [EventDraft] = []
        for (index, sub) in subEvents.enumerated() {
            guard let rawName = sub.name?.trimmingCharacters(in: .whitespacesAndNewlines), !rawName.isEmpty else { continue }
            guard let raw = sub.startDate, let startTimeUtc = Self.parseInstant(raw) else { continue }
            let label: String
            if let range = rawName.range(of: " - ") {
                let before = String(rawName[..<range.lowerBound]).trimmingCharacters(in: .whitespacesAndNewlines)
                label = before.isEmpty ? rawName : before
            } else {
                label = rawName
            }

            result.append(EventDraft(
                series: series,
                uid: "\(series.rawValue)-\(slug)-\(index)",
                fullTitle: "\(series.displayName) - \(roundName) - \(circuitName) - \(label)",
                startTimeUtc: startTimeUtc,
                timeZoneId: nil,
                inferredBadge: SessionBadgeMatcher.match(label)
            ))
        }
        return result
    }

    static func parseInstant(_ iso: String) -> Date? {
        let withFractional = ISO8601DateFormatter()
        withFractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = withFractional.date(from: iso) { return date }
        let standard = ISO8601DateFormatter()
        standard.formatOptions = [.withInternetDateTime]
        return standard.date(from: iso)
    }
}

/// Forma "evento con subEvent opcional" de schema.org/SportsEvent — también la reutiliza
/// `FormulaEScheduleSource` (mismo paquete en Android, mismo módulo aquí).
struct JsonLdSportsEvent: Decodable {
    let name: String?
    let startDate: String?
    let location: JsonLdLocation?
    let subEvent: [JsonLdSportsEvent]?
}

struct JsonLdLocation: Decodable {
    let name: String?
}

File: PitBoard(Android)/app/src/main/java/com/pitboard/app/schedule/sources/JsonLdSportsEventScheduleSource.kt
package com.pitboard.app.schedule.sources

import com.pitboard.app.data.EventEntity
import com.pitboard.app.data.RaceSeries
import com.pitboard.app.schedule.RaceScheduleSource
import com.pitboard.app.schedule.SessionBadgeMatcher
import com.pitboard.app.standings.StandingsHttpClient
import com.pitboard.app.standings.StandingsMoshi
import okhttp3.Request
import org.jsoup.Jsoup
import java.time.OffsetDateTime
import java.time.Year

/**
 * Fuente genérica para las webs que publican cada ronda con datos estructurados
 * `schema.org/SportsEvent` (bloque `<script type="application/ld+json">` con un array
 * `subEvent`, cada uno con `name` tipo "Practice - Australian Grand Prix" y `startDate` en
 * ISO-8601 con offset) — verificado a mano el 02/09/2026 en fiaformula2.com, fiaformula3.com
 * y europeanlemansseries.com, que comparten esta misma estructura pese a ser sitios distintos.
 *
 * Funciona en dos pasos:
 * 1. Se lee la página de listado de la temporada y se sacan los enlaces a cada ronda
 *    (empiezan por [roundHrefPrefixTemplate], con el año ya sustituido).
 * 2. Se visita cada ronda y se lee su JSON-LD.
 *
 * HONESTO: si el sitio cambia el marcado del listado o dejara de usar JSON-LD, esta fuente
 * empieza a devolver listas vacías (fallo silencioso y aislado, ver RaceScheduleRepository) —
 * no hay tabla HTML de la que caer hacia atrás en este caso.
 */
class JsonLdSportsEventScheduleSource(
    override val series: RaceSeries,
    private val baseUrl: String,
    /** URL de listado de la temporada, con "{year}" como marcador (ej.
     *  "https://www.fiaformula2.com/en/racing/{year}"). */
    private val listingUrlTemplate: String,
    /** Prefijo del href de cada ronda dentro del listado, con "{year}" como marcador (ej.
     *  "/en/racing/{year}/"). */
    private val roundHrefPrefixTemplate: String,
    /** Fragmentos que, si aparecen en el slug de la ronda, la descartan (ej. "test" para no
     *  colar los días de test oficiales en el calendario de carreras). */
    private val excludeSlugContaining: List<String> = emptyList()
) : RaceScheduleSource {

    override suspend fun fetch(): List<EventEntity> {
        val year = Year.now().value
        val listingUrl = listingUrlTemplate.replace("{year}", year.toString())
        val roundHrefPrefix = roundHrefPrefixTemplate.replace("{year}", year.toString())

        val listingHtml = fetchHtml(listingUrl)
        val roundUrls = extractRoundUrls(listingHtml, roundHrefPrefix)

        return roundUrls.flatMap { roundUrl -> runCatching { sessionsForRound(roundUrl) }.getOrElse { emptyList() } }
    }

    // 04/09/2026 (Fase 1 del diagnóstico): separado de fetch() para poder testear el filtro
    // de enlaces de ronda (prefijo + exclusión de slugs de test) contra un fixture HTML sin
    // red — ver JsonLdSportsEventScheduleSourceTest.
    internal fun extractRoundUrls(listingHtml: String, roundHrefPrefix: String): List<String> {
        val listingDoc = Jsoup.parse(listingHtml, baseUrl)
        return listingDoc.select("a[href]")
            .filter { it.attr("href").startsWith(roundHrefPrefix) }
            .mapNotNull { it.attr("abs:href").takeIf { url -> url.isNotBlank() } }
            .distinct()
            .filterNot { url -> excludeSlugContaining.any { it.lowercase() in url.lowercase() } }
    }

    private fun sessionsForRound(roundUrl: String): List<EventEntity> = parseRoundHtml(fetchHtml(roundUrl), roundUrl)

    // 04/09/2026 (Fase 1 del diagnóstico): separado de sessionsForRound() para poder
    // testear el parsing del JSON-LD de una ronda contra un fixture HTML sin red.
    internal fun parseRoundHtml(html: String, roundUrl: String): List<EventEntity> {
        val doc = Jsoup.parse(html, roundUrl)
        val adapter = StandingsMoshi.instance.adapter(JsonLdSportsEvent::class.java)

        val event = doc.select("script[type=application/ld+json]")
            .mapNotNull { script -> runCatching { adapter.fromJson(script.data()) }.getOrNull() }
            .firstOrNull { !it.subEvent.isNullOrEmpty() }
            ?: return emptyList()

        val slug = roundUrl.trimEnd('/').substringAfterLast('/')
        // El nombre del subEvent trae "Sesión - Nombre del Gran Premio" (ej. "Practice -
        // Australian Grand Prix") — la parte de después del guion es más corta y legible que
        // el "name" del evento raíz (que en fiaformula2/3.com viene con el título completo del
        // GP de F1, patrocinador incluido).
        val roundName = event.subEvent.orEmpty().firstNotNullOfOrNull { sub ->
            sub.name?.substringAfter(" - ", missingDelimiterValue = "")?.trim()?.takeIf { it.isNotBlank() }
        } ?: event.name?.trim().orEmpty()
        val circuitName = event.location?.name?.trim().orEmpty()

        return event.subEvent.orEmpty().mapIndexedNotNull { index, sub ->
            val rawName = sub.name?.trim() ?: return@mapIndexedNotNull null
            val startTimeUtc = sub.startDate?.let { parseInstant(it) } ?: return@mapIndexedNotNull null
            val label = rawName.substringBefore(" - ").trim().ifBlank { rawName }

            EventEntity(
                series = series,
                uid = "${series.name}-$slug-$index",
                fullTitle = "${series.displayName} - $roundName - $circuitName - $label",
                startTimeUtc = startTimeUtc,
                timeZoneId = null,
                inferredBadge = SessionBadgeMatcher.match(label)
            )
        }
    }

    private fun parseInstant(iso: String): Long? =
        runCatching { OffsetDateTime.parse(iso).toInstant().toEpochMilli() }.getOrNull()

    private fun fetchHtml(url: String): String {
        val request = Request.Builder()
            .url(url)
            .header("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36")
            .build()
        return StandingsHttpClient.instance.newCall(request).execute().use { response ->
            if (!response.isSuccessful) error("$url: HTTP ${response.code}")
            response.body?.string() ?: error("$url: cuerpo vacío")
        }
    }
}

data class JsonLdSportsEvent(
    val name: String?,
    val startDate: String?,
    val location: JsonLdLocation?,
    val subEvent: List<JsonLdSportsEvent>?
)

data class JsonLdLocation(val name: String?)
